### DISCUSSION TITLE
[Feedback wanted] Pattern for local atom implementation. Does it look good?

### DISCUSSION THREAD

**Discussion by wootra** [pmndrs/jotai#1951] [Q&A] [answered]
Hi, Team. 
I recently made a code like this:
```
export const useSelectedAtomValue = <T>(
	read: (get: Getter) => T,
	dependencies: React.DependencyList = [],
	atomOptions: Options | undefined = undefined
) => {
	// eslint-disable-next-line react-hooks/exhaustive-deps
	const anAtom = useMemo(() => atom(read), dependencies);
	return useAtomValue(anAtom, atomOptions);
};
// when using above, use it like this
// useSelectedAtomValue(get=>get(someAtom).someMap[id].data[slug], [id, slug], {store: myStore}); 
```
based on the suggestion. (https://github.com/pmndrs/jotai/discussions/1936#discussioncomment-5932570)
But I got a lot of anti-approval for the above code since that code does not work if I don't use useMemo.
from the react documentation, useMemo should be optional (https://react.dev/reference/react/useMemo#usage), but above code goes infinite loop since atom(read) will make always the new reference and useAtomValue use anAtom as a dependency.
<img width="915" alt="image" src="https://github.com/pmndrs/jotai/assets/6027633/a1d22f5d-aa68-4134-baa0-aaeeec63e57d">

Based on that, I used different approach using useRef instead, made below code:
```
export const useSelectedAtomValue = (
	read,
	dependencies = [],
	atomOptions: Options | undefined = undefined
) => {
	const dependenciesRef = useRef(null);
	const atomRef = useRef(null);
        // isSameArrayShallow is array shallow comparison function 
	if (!isSameArrayShallow(dependenciesRef.current, dependencies)) {
		atomRef.current = atom(read);
		dependenciesRef.current = dependencies;
	}
	const anAtom = atomRef.current!;
	return useAtomValue(anAtom, atomOptions);
};
```
How does it look like? the code passed test as first code passed.

**Comment by dai-shi** [comment]
`useMemo(() => atom((get) => ...), [...])` doesn't need a semantic guarantee, because it's just a derived atom.

It's close to useState and derived state:
```js
const [count, setCount] = useState(1);
const arr = useMemo(() => new Array(count), [count]);
```

The new React docs show typical use cases, but `useMemo` is a valid tool to get a stable reference.

On the other hand, because it can't be used as semantic guarantee, so creating a primitive atom is bad.
```js
useMemo(() => atom(1), []); // ❌ don't do this
```

**Comment by AjaxSolutions** [comment]
> useMemo(() => atom((get) => ...), [...]) doesn't need a semantic guarantee, because it's just a derived atom.

@dai-shi What will happen if React re-generates the atom.  This can happen even if the useMemo dependencies didn't change.

Won't the app will work differently, but it must work the same with or without the useMemo?

Also, could atomFamily be used in place of useMemo?

**Reply by dai-shi** [reply]
It works fine if React re-generates the atom, because it re-computes from other atoms.
useMemo is required to create different references in a **short** period.

**Comment by AjaxSolutions** [comment]
Does having a new atom config imply that a new value will be stored for that particular atom config?  If we use the new atom value as a dependency to a useEffect, then a new side effect may be created, i.e. the app will work differently.

Also, would using atomFamily instead of useMemo be a better option since atomFamily is supposed to cache atoms?

**Reply by dai-shi** [reply]
Yes, in that case, useRef or atomFamily would be better.

**Reply by wootra** [reply]
I considered atomFamily as well, but the parameter of the atomFamily can be dynamic value which can cause memory leak as atomFamily's doc mentions.  is it a right assumption?
```
const family = atomFamily((id,name)=>atom(get=>get(someAtom)[id][name]));
const Compo = ({id,name})=>{ // id, name can be dynamic. how should I remove dynamically?
  const atom = family(id,name);
  const value = useAtomValue(atom);
  useEffect(()=>{
    const saved = [id,name];
    return ()=>{
      family.delete(saved.id, saved.name); // we can remove it manually, but is it right place to put?
    }
  }, [id,name];
}

**Reply by wootra** [reply]
And I cannot understand the case can cause problem when the atom value is fed to a dependency of useEffect. and how useRef/atomFamily can fix the issue differently? can I have an example that possibly can go wrong? I tested multiple scenario, but couldn't find any case that possibly can render differently because of the new atom config. I think it should not since the new atom config will cause a new rendering with the new value. if it is given to useEffect, useEffect will cause another rendering.

**Reply by dai-shi** [reply, marked as answer]
For atomFamily, please see the simplified version and see why it can leak: https://github.com/pmndrs/jotai/discussions/1323#discussioncomment-3303035
It may not be a real issue depending on your use case.

For useEffect dependency, I think it's bad practice in the first place, and there should be a better workaround.

So, in summary, the answer again to your original question, `useMemo` for derived atoms, and `useRef` for primitive atoms.

**Reply by wootra** [reply]
thank you so much!
